### PR TITLE
Add logic for inferring newer broker versions

### PR DESCRIPTION
**Pull request by gabriel-tincu**
- New Fetch / ListOffsets request / response objects
- Add new test cases to inferr code based on mentioned objects
- Add unit test to check inferred version against whatever resides in KAFKA_VERSION
- Add new kafka broker versions to integration list
- Add more kafka broker versions to travis task list

As a side note, is there a ticket open or are there plans to use the same method the java client uses for keeping the Request/response classes up to date? It would seem much easier to import the folder with json api files and have our classes generated from that information, rather that doing a copy paste cycle multiple times, which can get very error prone

@jeffwidman

---
This change is [Reviewable](https://reviewable.io/reviews/dpkp/kafka-python/2038)

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -14,6 +14,10 @@ env:
     - KAFKA_VERSION=0.10.2.2
     - KAFKA_VERSION=0.11.0.3
     - KAFKA_VERSION=1.1.1
+    - KAFKA_VERSION=2.0.1
```
[the remark this answers is not in the document]
Comment (tvoinarovskyi) [reply .travis.yml:17]: Let's do only 2.5.0 and 2.4.0 for now.

File: kafka/conn.py [hunk 2]
```diff
@@ -1166,10 +1169,21 @@ def _infer_broker_version_from_api_versions(self, api_versions):
         # in reverse order. As soon as we find one that works, return it
         test_cases = [
             # format (<broker version>, <needed struct>)
+            ((2, 5, 0), DescribeAclsRequest[2]),
```
Comment (tvoinarovskyi) [inline kafka/conn.py:1172]: Great job on those versions. I had a hard time finding them, did you copy it from somewhere?))

File: build_integration.sh [hunk 3]
```diff
@@ -1,6 +1,6 @@
 #!/bin/bash
 
-: ${ALL_RELEASES:="0.8.2.2 0.9.0.1 0.10.1.1 0.10.2.2 0.11.0.3 1.0.2 1.1.1 2.0.1 2.1.1"}
+: ${ALL_RELEASES:="0.8.2.2 0.9.0.1 0.10.1.1 0.10.2.2 0.11.0.3 1.0.2 1.1.1 2.0.1 2.1.1 2.2.0 2.2.1 2.3.0 2.4.0 2.5.0"}
```
Comment (tvoinarovskyi) [inline build_integration.sh:3]: Let's stick to last minor versions, don't think there is a big difference between 2.2.0 and 2.2.1 in API...

**Comment by tvoinarovskyi** [comment]
LGTM, but let's not add too much to Travis, it runs really long.
@jeffwidman Seems fine to me, could you have a look through, don't want to merge without someone else looking )

**Comment by tvoinarovskyi** [comment]
@gabriel-tincu Great job!

**Comment by jeffwidman** [comment]
Thanks for working on this, I don't have time today but will try to look at this over the weekend.

File: kafka/conn.py (hunk 2 above)
Comment (gabriel-tincu) [reply kafka/conn.py:1172]: I ran the metadata request on each version of kafka, then plugged those in python and did an incremental diff. I was actually thinking of hardcoding those responses into the codebase as the actual api versions chosen would make more sense

File: .travis.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (gabriel-tincu) [reply .travis.yml]: fixed

File: build_integration.sh (hunk 3 above)
Comment (gabriel-tincu) [reply build_integration.sh]: fixed

File: kafka/conn.py [hunk 4]
```diff
@@ -1166,10 +1169,20 @@ def _infer_broker_version_from_api_versions(self, api_versions):
...
+            ((2, 3, 0), FetchRequest[11]),
+            ((2, 2, 0), OffsetRequest[5]),
+            ((2, 1, 0), FetchRequest[10]),
+            ((2, 0, 0), FetchRequest[8]),
+            ((1, 1, 0), FetchRequest[7]),
             ((1, 0, 0), MetadataRequest[5]),
             ((0, 11, 0), MetadataRequest[4]),
             ((0, 10, 2), OffsetFetchRequest[2]),
             ((0, 10, 1), MetadataRequest[2]),
+            # taken from https://github.com/edenhill/librdkafka/blob/master/src/rdkafka_feature.c#L234
```
[the remark this answers is not in the document]
Comment (gabriel-tincu) [reply kafka/conn.py]: fixed

File: kafka/conn.py [hunk 5]
```diff
@@ -24,9 +24,12 @@
 from kafka.future import Future
 from kafka.metrics.stats import Avg, Count, Max, Rate
 from kafka.oauth.abstract import AbstractTokenProvider
-from kafka.protocol.admin import SaslHandShakeRequest
+from kafka.protocol.admin import SaslHandShakeRequest, DescribeAclsRequest_v2
```
Comment (gabriel-tincu) [inline kafka/conn.py:27]: This plus my last commit is in response to the broker not recognizing the new V2 request even tho the schema is unchanged per 
https://github.com/apache/kafka/blob/2.5/clients/src/main/resources/common/message/DescribeAclsRequest.json
This leads me to believe that objects are serialized and deserialized differently when flexible versions are enabled , but i need some alone time with the java client and / or wireshark to figure out how that actually happens

Comment (tvoinarovskyi) [reply kafka/conn.py:27]: @gabriel-tincu Ohhhh, so they changed the protocol specification in version 2.4.0 basically. All new versions after that would require us to have support for Tagged fields on protocol level. For now, we can leave the auto-discovery up to 2.5, that's not a problem, but the change in Admin client above should be reverted if version 2 does not work.

Comment (tvoinarovskyi) [reply kafka/conn.py:27]: @gabriel-tincu As for the Compact versions of each type it should not be too hard, as we already have a varint implementation for length in protocol utils. It was used for message V2 parser.

Comment (gabriel-tincu) [reply kafka/conn.py:27]: reverted the admin code

**Comment by tvoinarovskyi** [comment]
@jeffwidman FYI. I will merge this PR with 1 minor point - it adds DescribeAclsRequest_v2 and DescribeAclsResponse_v2 that are not fully correct, as they don't abide to this new Protocol format, that supports Tagged fields. We need to add support to it before we can finish this class.

**Comment by tvoinarovskyi** [comment]
@gabriel-tincu Thanks greatly for the contributions!

**Comment by ofek** [comment]
Can this please be released? cc @dpkp

**Pull request merged**